### PR TITLE
[feat][minor] 2/3 Make it explicit whether an attention mechanism supports a mask

### PR DESCRIPTION
**Pull request by blefaudeux**
## What does this PR do?
Preamble to the Triton2 PR (since triton 2 will change the blocksparse attention and not support attention masks anymore)

The unit test which now fails on CI should be unrelated to this PR, and is fixed by the next PR in line, triton2

## Before submitting
- [x] Did you have fun?
  - Make sure you had fun coding 🙃
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/xformers/blob/master/CONTRIBUTING.md)?
- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [ ] N/A
- [x] Did you make sure to update the docs?
  - [ ] N/A
- [x] Did you write any new necessary tests?
  - [ ] N/A
- [x] Did you update the [changelog](https://github.com/facebookresearch/xformers/blob/master/CHANGELOG.md)? (if needed)
  - [ ] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -150,7 +150,7 @@ run_unittests: &run_unittests
   - run:
       name: Run Unit Tests
       command: |
-        pytest --junitxml=test-results/junit.xml --verbose --timeout 600 tests
+        CUDA_LAUNCH_BLOCKING=1 pytest --junitxml=test-results/junit.xml --verbose --timeout 600 tests
```
Comment (blefaudeux) [inline .circleci/config.yml:153]: @fmassa CI crash happens with or without this, I added that to make sure that it was not another test before which was silently failing. The crash is in a sputnik kernel, but it makes little sense to me given the contents of this PR

File: tests/test_model_factory.py [hunk 2]
```diff
@@ -39,7 +39,7 @@
         "num_heads": 4,
         "residual_dropout": 0,
         "attention": {
-            "name": "linformer",
+            "name": "scaled_dot_product",
```
Comment (blefaudeux) [inline tests/test_model_factory.py:42]: we were actually passing an attention mask in this test, which makes little sense with linformer. Now that we assert in this case, it had to be fixed :)

**Comment by blefaudeux** [comment]
@dianaml0 @finassa this goes with the triton2 PR again

File: xformers/components/attention/base.py [hunk 3]
```diff
@@ -53,6 +53,10 @@ def __init__(self, dropout: Optional[float] = None, *args, **kwargs):
         # so that the MHA wrapper should skip it
         self.requires_skip_multi_head = False
 
+        # Whether this attention mechanism supports attention masks
+        self.supports_attention_mask = True
+        self.supports_key_padding_mask = False
```
Comment (fmassa) [inline xformers/components/attention/base.py:58]: Ideally, I think I would prefer if we only support `attention_mask`, and then support the "optimized" case of `key_padding_mask` internally by checking the strides of the `attention_mask` Tensor.
The `key_padding_mask` case is actually just assuming that the stride of dimension -2 of the `attention_mask` is 0. So the converting `key_padding_mask` to a `attention_mask` can be done efficiently via
```python
attn_mask = key_padding_mask[:, None, None, :].expand(batch, heads, query_len, key_len)
```

Indeed, are we sure that all our implementations support both `attention_mask` *and* `key_padding_mask` at the same time?

**Review by fmassa — APPROVED** [review]
LGTM, thanks!

I have only one comment (which is sort of unrelated to this PR), but which I'd love to get your opinion on.

File: xformers/components/attention/base.py (hunk 3 above)
Comment (blefaudeux) [reply xformers/components/attention/base.py:58]: Agree on the simpler interface, it's a bit confusing right now. One trick is that for some attentions (Nystrom for instance) there's no attention mask but there's a possible key padding mask (which can then generate a mask of sorts indeed, but in that case the dimensions will differ). I think we could do a second pass on that indeed, thoughts @dianaml0  ?

Comment (dianaml0) [reply xformers/components/attention/base.py:58]: I agree ideally we could have this simpler approach but I ended up adding a key padding mask argument separately for Nystrom since Nystrom can't apply attention masks. I couldn't come up with a way around that but maybe there's a better way

**Review by dianaml0 — APPROVED** [review]

File: xformers/components/attention/base.py (hunk 3 above)
Comment (blefaudeux) [reply xformers/components/attention/base.py:58]: hmm so ideally we would just support key padding for nystrom and not make it visible everywhere ? right now there's a global flag because this happens in the MHA, but maybe that we could more this one level down and not expose this for other mechanisms ?

Comment (blefaudeux) [reply xformers/components/attention/base.py:58]: else it feels like an abstraction leak on our end, not perfect.. could be worth some more thinking
